File: src/Counsel/PlayerListViewController.swift
//
//  PlayerListViewController.swift
//  Counsel
//
//  Created by Han Hossain on 9/29/19.
//  Copyright © 2019 Han Hossain. All rights reserved.
//

import UIKit

class PlayerListViewController: UITableViewController {
    
    private let cellId = "playerListCell"
    private let fantasyService = FantasyService()
    
    private var players = [Player]()
    private var sections = [(offset: Int, initial: Character)]()

    override func viewDidLoad() {
        super.viewDidLoad()
        
        fantasyService.getCurrentWeek { (result) in
            let message: String
            
            switch (result) {
            case .success((let season, let week)):
                message = "Season: \(season) - Week: \(week)"
            case .failure(let error):
                print(error.localizedDescription)
                message = "Counsel"
            }
            
            DispatchQueue.main.async {
                self.title = message
            }
        }
        
        fantasyService.getPlayers { (result) in
            switch (result) {
            case .success(let players):
                self.savePlayers(players)
                
                DispatchQueue.main.async {
                    self.tableView.reloadData()
                }
            case .failure(let error):
                print(error.localizedDescription)
            }
        }
    }
    
    private func savePlayers(_ players: [String : Player]) {
        self.players = players.values
            .sorted(by: { (player1, player2) -> Bool in
                if player1.lastName.uppercased() == player2.lastName.uppercased() {
                    return player1.firstName.uppercased() < player2.firstName.uppercased()
                }
                
                return player1.lastName.uppercased() < player2.lastName.uppercased()
            })
        
        for (index, player) in self.players.enumerated() {
            if var playerInitial = player.lastName.first?.uppercased().first {
                if playerInitial.isNumber {
                    playerInitial = "#"
                }
                
                if sections.last?.initial != playerInitial {
                    sections.append((index, playerInitial))
                }
            }
        }
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return sections.count
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let sectionOffset = sections[section].offset
        let exclusiveUpperBound = section < sections.count - 1 ? sections[section + 1].offset : players.count
        return exclusiveUpperBound - sectionOffset
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: cellId, for: indexPath)
        
        let player = getPlayer(at: indexPath)
        cell.textLabel?.text = player.fullName
        cell.detailTextLabel?.text = "\(player.position) - \(player.team ?? "")"

        return cell
    }
    
    private func getPlayer(at indexPath: IndexPath) -> Player {
        let sectionOffset = sections[indexPath.section].offset
        let playerIndex = sectionOffset + indexPath.row
        return players[playerIndex]
    }
    
    override func sectionIndexTitles(for tableView: UITableView) -> [String]? {
        return sections.map { String($0.initial) }
    }

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}

File: src/Counsel.iOS/Controllers/PlayerListViewController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Counsel.Core;
using Counsel.Core.Models;
using Counsel.iOS.Views;
using Foundation;
using UIKit;

namespace Counsel.iOS.Controllers
{
	public sealed class PlayerListViewController : UITableViewController, ISearchDelegate
	{
		private const string _cellId = "playerListCell";

		private readonly IFantasyService _fantasyService;
		private readonly List<(char Initial, int Offset)> _sections = new List<(char, int)>();

		private List<Player> _players = new List<Player>();
		private UIBarButtonItem _resetButton;

		public PlayerListViewController(IFantasyService fantasyService)
		{
			_fantasyService = fantasyService;
		}

		public override async void ViewDidLoad()
		{
			base.ViewDidLoad();
			TableView.RegisterClassForCellReuse<SubtitleTableViewCell>(_cellId);

			var searchButton = new UIBarButtonItem(UIBarButtonSystemItem.Search);
			searchButton.Clicked += SearchButton_Clicked;

			NavigationItem.RightBarButtonItem = searchButton;

			_resetButton = new UIBarButtonItem()
			{
				Title = "Reset"
			};
			_resetButton.Clicked += async (s, e) =>
			{
				await LoadAllPlayersAsync();
				InvokeOnMainThread(() => TableView.ReloadData());
			};

			var loadingAlert = UIAlertController.Create("Loading...", null, UIAlertControllerStyle.Alert);
			InvokeOnMainThread(() => PresentViewController(loadingAlert, true, null));

			await _fantasyService.UpdateAsync();
			await LoadCurrentWeekAsync();
			await LoadAllPlayersAsync();

			InvokeOnMainThread(() =>
			{
				DismissViewController(true, null);
				TableView.ReloadData();
			});
		}

		public override nint NumberOfSections(UITableView tableView)
		{
			return _sections.Count;
		}

		public override nint RowsInSection(UITableView tableView, nint section)
		{
			if (!_sections.Any())
			{
				return 0;
			}

			int sectionOffset = _sections[(int)section].Offset;
			int exclusiveUpperBound = section < _sections.Count - 1
				? _sections[(int)section + 1].Offset
				: _players.Count;

			return exclusiveUpperBound - sectionOffset;
		}

		public override string[] SectionIndexTitles(UITableView tableView)
		{
			return _sections
				.Select(x => x.Initial.ToString())
				.ToArray();
		}

		public override UITableViewCell GetCell(UITableView tableView, NSIndexPath indexPath)
		{
			var cell = tableView.DequeueReusableCell<SubtitleTableViewCell>(_cellId, indexPath);
			var player = GetPlayer(indexPath);

			cell.TextLabel.Text = player.FullName;
			cell.DetailTextLabel.Text = $"{player.Position} - {player.Team}";
			return cell;
		}

		public override void RowSelected(UITableView tableView, NSIndexPath indexPath)
		{
			var player = GetPlayer(indexPath);
			var playerDetailViewController = new PlayerDetailViewController(player, _fantasyService);
			NavigationController.PushViewController(playerDetailViewController, true);
		}

		private Player GetPlayer(NSIndexPath indexPath)
		{
			int playerIndex = _sections[indexPath.Section].Offset + indexPath.Row;
			var player = _players[playerIndex];
			return player;
		}

		private async Task LoadCurrentWeekAsync()
		{
			(int season, int week) = await _fantasyService.GetCurrentWeekAsync();
			BeginInvokeOnMainThread(() => Title = $"Season: {season} - Week: {week}");
		}

		private async Task LoadAllPlayersAsync()
		{
			var players = await _fantasyService.GetPlayersAsync();
			LoadPlayers(players);

			InvokeOnMainThread(() => NavigationItem.SetLeftBarButtonItem(null, true));
		}

		private void LoadPlayers(Dictionary<string, Player> players)
		{
			_players = players.Values
				.OrderBy(x => x.LastName)
				.ThenBy(x => x.FirstName)
				.ToList();

			LoadSections();
		}

		private void LoadSections()
		{
			_sections.Clear();

			for (int i = 0; i < _players.Count; i++)
			{
				var player = _players[i];
				char playerInitial = player.LastName.First();

				if (char.IsNumber(playerInitial))
				{
					playerInitial = '#';
				}

				if (_sections.Count == 0 || playerInitial != _sections[_sections.Count - 1].Initial)
				{
					_sections.Add((playerInitial, i));
				}
			}
		}

		private void SearchButton_Clicked(object sender, EventArgs e)
		{
			var navController = new UINavigationController(new SearchViewController(this));
			PresentViewController(navController, true, null);
		}

		public async Task OnSearchAsync(string playerName)
		{
			var players = await _fantasyService.SearchPlayersAsync(playerName);
			LoadPlayers(players);

			InvokeOnMainThread(() =>
			{
				NavigationItem.SetLeftBarButtonItem(_resetButton, true);
				TableView.ReloadData();
			});
		}
	}
}
